x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -3,14 +3,821 @@
+            // the metadata blob container mirrors its structure, but in some failure cases it might contain
+            // additional orphaned metadata blobs
+            Map<String, BlobContainer> childEncryptedBlobContainers = delegatedBlobContainer.children();
+            Map<String, BlobContainer> result = new HashMap<>(childEncryptedBlobContainers.size());
+            for (Map.Entry<String, BlobContainer> encryptedBlobContainer : childEncryptedBlobContainers.entrySet()) {
+                if (encryptedBlobContainer.getValue().path().equals(encryptionMetadataBlobContainer.path())) {
+                    // do not descend recursively into the metadata blob container itself
+                    continue;
+                }
+                // get an encrypted blob container for each child
+                // Note that the encryption metadata blob container might be missing
+                result.put(encryptedBlobContainer.getKey(), new EncryptedBlobContainer(delegatedBlobStore, delegatedBasePath,
+                        path.add(encryptedBlobContainer.getKey()), dataEncryptionKeySupplier, metadataEncryption,
+                        encryptionNonceSupplier, metadataIdentifierSupplier));
+            }
+            return result;
+        }
+    }
+
+    private static String computeSaltedPBKDF2Hash(SecureRandom secureRandom, char[] password) {
+        byte[] salt = new byte[PASSWORD_HASH_SALT_LENGTH_IN_BYES];
+        secureRandom.nextBytes(salt);
+        return computeSaltedPBKDF2Hash(salt, password);
+    }
+
+    private static String computeSaltedPBKDF2Hash(byte[] salt, char[] password) {
+        final PBEKeySpec spec = new PBEKeySpec(password, salt, SALTED_PASSWORD_HASH_ITER_COUNT, SALTED_PASSWORD_HASH_KEY_LENGTH_IN_BITS);
+        final byte[] hash;
+        try {
+            SecretKeyFactory pbkdf2KeyFactory = SecretKeyFactory.getInstance(SALTED_PASSWORD_HASH_ALGO);
+            hash = pbkdf2KeyFactory.generateSecret(spec).getEncoded();
+        } catch (InvalidKeySpecException | NoSuchAlgorithmException e) {
+            throw new RuntimeException("Unexpected exception when computing the hash of the repository password", e);
+        }
+        return new String(Base64.getUrlEncoder().withoutPadding().encode(salt), StandardCharsets.UTF_8) + ":" +
+                new String(Base64.getUrlEncoder().withoutPadding().encode(hash), StandardCharsets.UTF_8);
+    }
+
+    /**
+     * Called before every snapshot operation on every node to validate that the snapshot metadata contains a password hash
+     * that matches up with the repository password on the local node.
+     *
+     * @param snapshotUserMetadata the snapshot metadata to verify
+     * @param exception the exception handler to call when the repository password check fails
+     */
+    private void validateRepositoryPasswordHash(Map<String, Object> snapshotUserMetadata, Consumer<Exception> exception) {
```

Comment:
I think this is a somewhat weird API. If verification fails we resolve the listener with an exception in the calling code, but still keep going with the rest of e.g. `finalizeSnapshot`.  Maybe it's easier to just make this method throw and then `try {} catch {}` in the calling code or make it return `IllegalStateException` or `null` (if all is well) and handle accordingly in the callers?

Reply:
Yes, this was wrong ! Fixed in 262b3e9